docs/src/lucet-runtime/killswitch.md

```diff
@@ -0,0 +1,346 @@
+select, we can discuss the actual lifecycle of the `KillState` in a Lucet
+`Instance` and see how these pieces begin to fit together.
+
+A Lucet `Instance` begins with a default `KillState`: in the `Pending` domain,
+`terminable` set to `true`, and no `thread_id` as it's not currently running.
+When a call is made to `Instance::run`, or more specifically,
+`Instance::run_func`, the `Instance`'s bootstrap context is set up and
+`KillState::schedule` is called to set up last-minute state before the instance
+begins running. In implementation, the only state to set up is the instance's
+thread ID - we assume that the instance won't be migrated between threads
+without `lucet_runtime`'s cooperation, so the thread ID at this point can be
+safely retained.
+
+`lucet_runtime` shortly thereafter will switch to the `Instance` and begin
+executing its bootstrapping code. In `enter_guest_region` the guest will lock
+and update the execution domain to `Guest`, or see the instance is
+`Domain::Cancelled` and exit. If the instance could run, we proceed into the
+AOT-compiled wasm. At some point, the instance will likely make a hostcall to
+embedder-provided or `lucet_runtime`-provided code; correctly-implemented
+hostcalls are wrapped in `begin_hostcall` and `end_hostcall` to take care of
+runtime bookkeeping including updating `execution_domain` to `Domain::Hostcall`
+(`begin_hostcall`) and checking that the instance can return to guest code,
+setting `execution_domain` back to `Domain::Guest` (`end_hostcall`). At some
+point the instance will hopefully exit, where in `lucet_context_backstop` we
+check that the instance may stop exiting (`KillState::terminable`). If it can
+exit, then do so. Finally, back in `lucet_runtime`, we can
+`KillState::deschedule` to tear down the last of the run-specific state - the
+`thread_id`.
+
+## Implementation Complexities (When You Have A Scheduler Full Of Demons)
+
+Many devils live in the details. The rest of this chapter will discuss the
+numerous edge cases and implementation concerns that Lucet's asynchronous
+signal implementation must consider, and arguments for its correctness in the
+face of these.
+
+First, a flow chart of the various states and their transitions:
+![state flow chart](states.png)
+
+This graph describes the various states that reflect values of
+`execution_domain`, and includes some timeout transitions reflecting a timeout
+occurring entirely in a specific domain, well away from racing with other
+machinery in wasm guest. The rest of this section discusses the correcness of
+timeouts at boundaries where these state transitions occur.
+
+For reference later, the possible state transitions are:
+* `A -> B`
+* `A -> D`
+* `B -> C`
+* `B -> E` (normal guest exit)
+* `B -> E` (from a guest fault/timeout)
+* `C -> B`
+* `C -> E` (hostcall terminates instance)
+* `C -> E` (hostcall observes timeout)
+  - not an internal state but we will also discuss timeouts during a hostcall fault
+* `D -> E`
```

Comment:
I might add descriptive notes to the other cases as well, in addition to those already noted. For example, `A -> B` means the guest started. `A -> D` is a cancelled guest, and so on.

I'd add these to the headers below too, so that somebody doesn't need to reference the diagram to understand what these transitions mean.

Reply:
Plus 1.

Reply:
I was afraid that copying the descriptions on each header would be too repetitive, but looking again after a few days away.. it's definitely good for the reader :grin: